### PR TITLE
[mesheryctl] Add skip-update flag on `system restart` cmd

### LINKED ISSUE #2906: [mesheryctl] Add `--skip-update` flag to `system restart` command
_See [mesheryctl Command Tracker](https://bit.ly/3dqXy1q) for current status of commands._

#### Current Behavior
`system start` offers user the ability the flag `--skip-update`, which has the effect of allowing them to bypass checking for and downloading new versions of Meshery's container images.

#### Desired Behavior
Ideally, `system restart` supports the same `--skip-update` flag with the same behavior.

---

#### Contributor Resources

- [mesheryctl Contributing Guide](https://github.com/layer5io/meshery/blob/master/mesheryctl/README.md)
- [Beginner's guide to contributing to Meshery and mesheryctl](https://youtu.be/hh_kFLZx3G4)
- [mesheryctl Command Tracker](https://docs.google.com/spreadsheets/d/1q63sIGAuCnIeDs8PeM-0BAkNj8BBgPUXhLbe1Y-318o/edit#gid=0)
- [Meshery CLI Commands and Documentation](https://docs.google.com/document/d/1xRlFpElRmybJ3WacgPKXgCSiQ2poJl3iCCV1dAalf0k/edit#heading=h.5fucij4hc5wt)

### PR DESCRIPTION
**Pull request by alphaX86**
Signed-off-by: Aadhitya A <aadhitya864@gmail.com>

**Description**

This PR fixes #2906 

**Notes for Reviewers**
Tested locally and it worked. Maintainers do check if it needs any changes!

**[Signed commits](https://github.com/layer5io/meshery/blob/master/CONTRIBUTING.md#signing-off-on-commits-developer-certificate-of-origin)**
- [x] Yes, I signed my commits.

### CODE REVIEW

**Comment by alphaX86** [comment]
Ugh... Need to lint the files...

**Comment by metonymic-smokey** [comment]
@alphaX86 I tried running this without starting Meshery in the first place. It says Meshery is restarted but running `system status`/`system stop` shows Meshery is not running.    
image

image

Maybe this could be a condition you could check for? Check if Meshery is not running and start it in that case.

**Comment by alphaX86** [comment]
@metonymic-smokey 😮 I didn't encounter this actually yet. Anyway, I'll check it though! 👍

**Comment by alphaX86** [comment]
@metonymic-smokey I tried your way
image

I didn't encounter error actually. I stopped in middle to show the output

**Comment by alphaX86** [comment]
And this is when meshery is running...
image

**Comment by alphaX86** [comment]
IDK what went wrong then... 🤔

**Comment by alphaX86** [comment]
@leecalcote what do you think about this? Do I need to dig further or ask any comments from reviewers?

File: mesheryctl/internal/cli/root/system/restart.go [hunk 1]
```diff
@@ -66,6 +66,14 @@ func restart() error {
...
 		}
 
+		if skipUpdateFlag {
+			log.Info("Skipping Meshery Update...")
+		} else {
+			if err := start(); err != nil {
+				return errors.Wrap(err, utils.SystemError("Failed to restart Meshery"))
+			}
+		}
+
```
Comment (navendu-pottekkat) [inline mesheryctl/internal/cli/root/system/restart.go:76]: In the restart command, I guess we need to only set the `skipUpdateFlag` as the actual function is carried out by `start()`. 

There is already a logging in `start.go` that does this, so no additional logic needs to be written here.

Please check line 178 in `start.go` and confirm-

```go
		if skipUpdateFlag {
			log.Info("Skipping Meshery update...")
		} else {
			err := utils.UpdateMesheryContainers()
			if err != nil {
				return errors.Wrap(err, utils.SystemError("failed to update Meshery containers"))
			}
		}
```

File: mesheryctl/internal/cli/root/system/restart.go [hunk 2]
```diff
@@ -97,3 +105,7 @@ func restart() error {
 	}
 	return nil
 }
+
+func init() {
+	restartCmd.Flags().BoolVarP(&skipUpdateFlag, "skip-update", "", false, "(optional) skip checking for new Meshery's container images.")
```
Comment (navendu-pottekkat) [inline mesheryctl/internal/cli/root/system/restart.go:110]: We would only need to register the flag here I guess and it is centrally controlled in `start.go` for all the system commands.

**Review by navendu-pottekkat — CHANGES_REQUESTED** [review]
@alphaX86 Please check the suggestions.

On a related note, for `system start --skip-update` `platform: kubernetes` does not take the flag into account. Need to work out how it will handle that.

// @metonymic-smokey We are finding bugs already

**Comment by navendu-pottekkat** [comment]
@metonymic-smokey Yes, this command is also missing a `isMesheryRunning` check.

**Comment by navendu-pottekkat** [comment]
Or, if Meshery is not running, it can automatically just start it. Maybe providing that as a log to the user

File: mesheryctl/internal/cli/root/system/restart.go (hunk 2 above)
Comment (alphaX86) [reply mesheryctl/internal/cli/root/system/restart.go:110]: Got it. But won't we need to print during help flag used?

File: mesheryctl/internal/cli/root/system/restart.go (hunk 1 above)
Comment (alphaX86) [reply mesheryctl/internal/cli/root/system/restart.go:76]: Alright

Comment (alphaX86) [reply mesheryctl/internal/cli/root/system/restart.go:76]: So, instead of re-writing the whole fn again, what I need to do in such case?

Comment (alphaX86) [reply mesheryctl/internal/cli/root/system/restart.go:76]: @navendu-pottekkat oh yeah. I removed the fn just now and it worked! 😨

File: mesheryctl/internal/cli/root/system/restart.go (hunk 2 above)
Comment (alphaX86) [reply mesheryctl/internal/cli/root/system/restart.go:110]: I mean when --help is used....

**Comment by alphaX86** [comment]
> @alphaX86 Please check the suggestions.
> 
> On a related note, for `system start --skip-update` `platform: kubernetes` does not take the flag into account. Need to work out how it will handle that.
> 
> // @metonymic-smokey We are finding bugs already

I didn't consider `kubernetes` for now. I also saw the source code `start.go` seems it applies only on Docker for now

File: mesheryctl/internal/cli/root/system/restart.go (hunk 1 above)
Comment (piyushsingariya) [reply mesheryctl/internal/cli/root/system/restart.go]: @alphaX86 What Navendu is saying here is to only set the value of `skipUpdateFlag` to be true with the help of flag, That's all. That'll handle the whole logic.

File: mesheryctl/internal/cli/root/system/restart.go (hunk 2 above)
Comment (piyushsingariya) [reply mesheryctl/internal/cli/root/system/restart.go:102]: That's all. Only this was needed.

File: mesheryctl/internal/cli/root/system/restart.go (hunk 1 above)
Comment (alphaX86) [reply mesheryctl/internal/cli/root/system/restart.go]: Oooh OK ok

File: mesheryctl/internal/cli/root/system/restart.go (hunk 2 above)
Comment (alphaX86) [reply mesheryctl/internal/cli/root/system/restart.go:102]: Interesting....

File: mesheryctl/internal/cli/root/system/restart.go (hunk 1 above)
Comment (alphaX86) [reply mesheryctl/internal/cli/root/system/restart.go]: I'll change it after lunch

File: mesheryctl/internal/cli/root/system/restart.go [hunk 3]
```diff
@@ -97,3 +97,7 @@ func restart() error {
 	}
 	return nil
 }
+
```
Comment (piyushsingariya) [inline mesheryctl/internal/cli/root/system/restart.go:100]: It works correctly on the Docker platform, there's not a need for `skip-update` for Kubernetes in `system restart`. 
Nice work, @alphaX86

File: mesheryctl/internal/cli/root/system/restart.go [hunk 4]
```diff
@@ -97,3 +97,7 @@ func restart() error {
...
+func init() {
+	restartCmd.Flags().BoolVarP(&skipUpdateFlag, "skip-update", "", false, "(optional) skip checking for new Meshery's container images.")
```
Comment (piyushsingariya) [inline mesheryctl/internal/cli/root/system/restart.go:102]: It works correctly on the Docker platform, there's not a need for `skip-update` for Kubernetes in `system restart`. 
LGTM, Nice work! @alphaX86

**Review by piyushsingariya — APPROVED** [review]

**Comment by metonymic-smokey** [comment]
> Or, if Meshery is not running, it can automatically just start it. Maybe providing that as a log to the user

Yes, that's what I was thinking too.

For Kubernetes, we'll have to find some way to download the latest manifests as updates, maybe by comparing file hashes or something similar.

**Comment by piyushsingariya** [comment]
> For Kubernetes, we'll have to find some way to download the latest manifests as updates, maybe by comparing file hashes or something similar.

@metonymic-smokey We just need to trigger `start` if meshery isn't running, right?

**Comment by metonymic-smokey** [comment]
> > For Kubernetes, we'll have to find some way to download the latest manifests as updates, maybe by comparing file hashes or something similar.
> 
> @metonymic-smokey We just need to trigger `start` if meshery isn't running, right?

If the `skip-update` flag has been used, yes.  
If the flag has been used, we don't download the manifests again.

**Comment by alphaX86** [comment]
@metonymic-smokey @piyushsingariya IG we can talk about this during the bug hunt meet today. What do you say guys?

**Comment by alphaX86** [comment]
Meanwhile, I'll make some minor changes in my PR, test them, lint the file and I'll send it in upstream

**Comment by piyushsingariya** [comment]
> I was thinking of a way to avoid overwriting and downloading manifests every time we need to update.

@metonymic-smokey I guess a PR is in progress to use cached manifests, we might be able to use that for this. #2807

**Comment by metonymic-smokey** [comment]
Yes, let's do that.
I'll add it to the test plan.

**Comment by alphaX86** [comment]
> Meanwhile, I'll make some minor changes in my PR, test them, lint the file and I'll send it in upstream

Rechecked again, seems I need not change again. Reviewers test the change and see if it works good

**Comment by alphaX86** [comment]
Any reviewer here? 😅

**Comment by piyushsingariya** [comment]
Yeah, me!

**Review by leecalcote — APPROVED** [review]
Looks good. Works good.

**Pull request merged**